File: C#/230. Kth Smallest Element in a BST.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution
{
    private int GetNum(TreeNode node, int k, out bool flag)
    {
        if (node == null)
        {
            flag = false;
            return 0;
        }
        if (k == 1)
        {
            flag = true;
            return node.val;
        }
        int nl = GetNum(node.left, k - 1, out flag);
        if (flag)
            return nl;
        if (nl + 1 == k)
        {
            flag = true;
            return node.val;
        }
        int nr = GetNum(node.right, k - nl - 1, out flag);
        if (flag)
            return nr;
        return nl + nr + 1;
    }
    public int KthSmallest(TreeNode root, int k)
    {
        if (root.left == null && root.right == null)
            return root.val;
        bool flag;
        int n = GetNum(root.left, k, out flag);
        if (flag)
            return n;
        if (k - n - 1 == 0)
            return root.val;
        return KthSmallest(root.right, k - n - 1);
    }
}

File: python/230. Kth Smallest Element in a BST.py
# -*- coding: utf-8 -*-

"""
@Author: xiezizhe
@Date: 24/6/2020 上午2:11
"""


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def dfs(self, node, k) -> (TreeNode, int):
        if not node:
            return None, 0

        if not node.left and not node.right:
            if k <= 1:
                return node, 1
            else:
                return None, 1

        lnode, lsize = self.dfs(node.left, k)
        if lnode:
            return lnode, -1

        if lsize + 1 == k:
            return node, -1

        rnode, rsize = self.dfs(node.right, k - lsize - 1)
        if rnode:
            return rnode, -1

        return None, lsize + rsize + 1

    def kthSmallest(self, root: TreeNode, k: int) -> int:
        node, _ = self.dfs(root, k)
        return node.val
